### ISSUE TITLE
Non-string keys

### ISSUE THREAD

**Issue by singpolyma** [ipld/specs#58]
Are non-string keys allowed in maps? When I try with current `ipfs dag put` it indicates no, but wondering if that is current implementation or a design from the spec.

**Comment by vmx** [comment]
I don't fully understand what you mean. Can you please give an example?

**Comment by singpolyma** [comment]
Easiest example would be keys that are integers

**Comment by vmx** [comment]
IPLD is format agnostic. I guess you are using talking about serializing maps to CBOR? From a quick look at the CBOR spec it seems it supports any binary data as keys.

If your input data is JSON, they keys can only be strings according to the JSON spec.

**Comment by singpolyma** [comment]
Yes, using IPLD's native format, CBOR, when I use integers as keys and feed to `ipfs dag put` I get the error that my keys are not strings.

**Comment by vmx** [comment]
Thanks for the info. I think I can finally answer your original question. It is the implemention. i think the CBOR serialization should support all valid CBOR.

@Stebalien Do you agree?

**Comment by Stebalien** [comment]
TL;DR: No, they're not valid (at the moment, as far as I know).

In order to *path* through an object, the key needs to be a string. While I don't think any of our current specs/documents say anything conclusive on the subject, I believe all current implementations expect the map keys to be strings (go does, at least).

The original intention was to actually be quite restrictive: map keys must be unicode strings with no slashes. However, we've loosened that so that they *can* contain slashes, it's just that those keys can't be pathed through. At this point, we may want to consider loosening it further but... it's tricky. Being able to path through an object was a highly desired property.

**Comment by vmx** [comment]
I didn't know that go-ipfs is validating the input of `dag put` even when put in as `raw`. That's good to know.

**Comment by singpolyma** [comment]
> map keys must be unicode strings with no slashes

That sounds totally reasonable to me, otherwise you need some sort of encoding mechanism to be able to path-address (which is a big part of the whole point) -- but whatever the restriction is going to be should make it into the spec, IMHO

**Comment by vmx** [comment]
@singpolyma You are right. I've created issue https://github.com/ipld/specs/issues/59 for that.

**Comment by Stebalien** [comment]
Thinking about this a bit, the important part is really being able to name any part of an IPLD datastructure (like how one can create a pointer to any part of a struct). Given that, supporting non-string keys is probably reasonable as long as we can make pathing unambiguous. That means:

* The keys must all be of the same type: `/0/` through `"0"` and `0` is ambiguous. CBOR recommends this anyways.
* The key type must define a unique, lossless mapping to strings.

Motivations:

1. Binary keys are useful in *many* datastructures.
2. Sparse arrays are often represented using maps.

**Comment by mikeal** [comment]
Per data model, no non-string keys in data model maps since you can’t reliably implement them in every language.

**Comment by singpolyma** [comment]
>Per data model, no non-string keys in data model maps since you can’t reliably implement them in every language.

That seems... false? Is there actually a language that cannot represent
bytestrings or integers at all?  I mean, I guess shell can't but then you could
use a base encoding for such weird edge cases, many shell scripts end up doing
just that for that reason.

**Comment by mikeal** [comment]
I should have been more specific. The default map type (or object type) in some languages only support string keys. JavaScript is the most obvious one (and has the largest market share of any language) so it’s problematic to allow this.

This is also why JSON only allows string keys in maps.

**Comment by singpolyma** [comment]
>I should have been more specific. The default map type (or object type) in 
>some languages only support string keys. JavaScript is the most obvious one 
>(and has the largest market share of any language) so it’s problematic to 
>allow this.

JavaScript CBOR libraries could use a custom map-compatible object type, no?

**Comment by mikeal** [comment]
The language doesn’t really allow that level of object customization. You can sort of use proxy objects for this but they are slow and don’t always behave natively. It just isn’t what people working with data actually want.

**Comment by vmx** [comment]
You could implement your own custom class that has Object like properties, like being iterable etc. In other languages we also have custom IPLD types. JavaScript is a special case here (and hopefully we can get away with native types).

**Comment by mikeal** [comment]
> You could implement your own custom class that has Object like properties

You can’t capture the `set()` operation for any new property being set using a custom class. The only way to do that is to use a Proxy object which is going to be slow and quite problematic in terms of representation.

We need the data model to cast into common and widely accessible native types, which means string-only keys. There’s numerous threads about this, but we just don’t have a reliable cross-language way to represent non-string keys in maps. Even in languages where it is supported, it’s often not ideal and can lead to performance issues.

We can go back and forth all day about how someone might find a less than ideal JavaScript representation, but whatever you end up with is far from the common usage, which means that it just won’t integrate nicely into everyone’s libraries and frameworks and won’t be future proof. As an example, we probably want to leverage the upcoming immutable types https://github.com/tc39/proposal-record-tuple and the map type (`record`) **also** only supports string keys.

On the other hand, if we just stay close to JSON’s constraints we know that we can find a well laid path to support in different languages because of JSON’s ubiquity.

If you’re in a situation in which it’s highly preferable to use non-string keys then you can easily create your own data structure that serializes to a list of key/value pairs. This is a very important point: **the data model does not limit what you can build as a data structure in your language, it only constrains how you need to serialize it.** The data model is not the end of the stack, it’s just a set of constraints specific to problems related to serialization, once you’re taking the de-serialized data model and working with it in your language you can do whatever you want and create numerous custom types.

It’s very common to serialize maps to lists of key/value pairs. The JavaScript `Map` type does this by default since it can’t serialize to a native object as it supports other key types ;)
